Amazon Top 50 Bestselling Books 2009 - 2019



https://drive.google.com/file/d/17_itPPjDnTeZ-KoEpvsp-0m2zHg6F7df/view?usp=sharing



In [ ]:
! gdown --id 17_itPPjDnTeZ-KoEpvsp-0m2zHg6F7df

/usr/local/lib/python3.10/dist-packages/gdown/__main__.py:132: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From: https://drive.google.com/uc?id=17_itPPjDnTeZ-KoEpvsp-0m2zHg6F7df
To: /content/bestsellers with categories.csv
100% 51.2k/51.2k [00:00<00:00, 62.8MB/s]


In [ ]:
import pandas as pd
df = pd.read_csv('bestsellers with categories.csv')


In [ ]:
df


Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[550 rows x 7 columns]

Problems:

1)Find out which books became popular more often.

2)How does the rating differ between fiction and non-fiction texts?

3)Are the average book ratings different for the last three years?(2017-2019)


Problem №1

Find out which books became popular more often.

To solve this poblem it can be used Binomial test

H0: The probability of finding a fiction and non-fiction book in our selection is equal

H1: The probability of finding a fiction and non-fiction book in our selection isn't equal

The level of significance  a = 0.05

In [ ]:
from scipy.stats import binom_test

In [ ]:
bins = df['Genre'].value_counts().tolist()

In [ ]:
print(bins)

[310, 240]


In [ ]:
binom_test(bins)

<ipython-input-7-e1b6044464f4>:1: DeprecationWarning: 'binom_test' is deprecated in favour of 'binomtest' from version 1.7.0 and will be removed in Scipy 1.12.0.
  binom_test(bins)


0.0032232676763761965

p-valeu = 0.00322

It means that p-valeu < a (a level of significance)

We can reject the H0:The probability of finding a fiction and non-fiction book in our selection is equal


Problem №2

How does the rating differ between fiction and non-fiction texts?

To solve this poblem it can be used T-test

H0: There is no significant difference between the average ratings of fiction and non-fiction books

H1:There is significant difference between the average ratings of fiction and non-fiction books

The level of significance a = 0.05



We will conduct a t-test to determine if there are significant differences between the ratings of fiction and non-fiction books, testing the null hypothesis of no such differences. First, we will check the normality of the distribution of average ratings using the Shapiro-Wilk test with 500 random samples. For each sample, we will calculate the average rating and record the results in a list.



In [14]:
means_fic = []
n = 1

while n < 500:
  n += 1
  sample = df[df['Genre'] == "Non Fiction"].sample(100)['User Rating'].mean()
  means_fic.append(sample)

In [15]:
from scipy.stats import shapiro
shapiro(means_fic)


ShapiroResult(statistic=0.9978311657905579, pvalue=0.7770273685455322)

The p-value is not significant, so we cannot reject the null hypothesis and assume that the means in the samples are normally distributed.

Next, we need to check the equality of variances between the groups to adjust a parameter for the t-test, using the levene function from the scipy package. For this, we will again create two random samples for comparison.

In [13]:
from scipy.stats import levene
user_rating_fic = df[df['Genre'] == "Fiction"].sample(100)['User Rating']
user_rating_non = df[df['Genre'] == "Non Fiction"].sample(100)['User Rating']
levene(user_rating_fic, user_rating_non)


LeveneResult(statistic=4.315605928509152, pvalue=0.039054415451977104)

We observe that the p-value is large, indicating that we accept the null hypothesis of equal variances. We need to reflect this in the t-test.

In [11]:
from scipy.stats import ttest_ind
ttest_ind(user_rating_fic, user_rating_non, equal_var=True)


TtestResult(statistic=3.0185478310958636, pvalue=0.002873947162869027, df=198.0)

Since variances are equal (using equal_var), and with a p-value less than 0.05, we reject the null hypothesis that book ratings do not vary by genre, indicating genre affects book ratings.

Problem №3

Are the average book ratings different for the last three years?(2017-2019)

Let's test if average book ratings differ over the last three years using an ANOVA test.

H0:The book ratings was the same over this period.

H1:The book ratings wasn't the same over this period.

a = 0.05


In [16]:
from scipy.stats import f_oneway
ur_2019 = df[(df['Year'] == 2019)]['User Rating']
ur_2018 = df[(df['Year'] == 2018)]['User Rating']
ur_2017 = df[(df['Year'] == 2017)]['User Rating']
f_oneway(ur_2019, ur_2018, ur_2017)

F_onewayResult(statistic=2.99211541687636, pvalue=0.05324686656835848)


Based on the p-value threshold (0.053 > 0.05), we cannot reject the null hypothesis. This indicates that the ratings of top books throughout the year do not differ from each other.